### ISSUE TITLE
@grpc/grpc-js is incompatible with typescript definations @types/node@17.0.6

### ISSUE THREAD

**Issue by hywhuangyuwei** [grpc/grpc-node#2002]
### Problem description
when project already has the dependency `@types/node@17.0.6`, `tsc` failed to build

### Reproduction steps

- `yarn add @types/node@17.0.6 && tsc -b`
- failed with output: 
```
node_modules/@grpc/grpc-js/build/src/object-stream.d.ts:11:18` - error TS2430: Interface 'IntermediateObjectWritable<T>' incorrectly extends interface 'Writable'.
  The types returned by 'end(...)' are incompatible between these types.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

11 export interface IntermediateObjectWritable<T> extends Writable {
                    ~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@grpc/grpc-js/build/src/server-call.d.ts:64:5 - error TS2416: Property 'end' in type 'ServerWritableStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Writable'.
  Type '(metadata?: any) => void' is not assignable to type '{ (cb?: () => void): this; (chunk: any, cb?: () => void): this; (chunk: any, encoding: BufferEncoding, cb?: () => void): this; }'.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

64     end(metadata?: any): void;
       ~~~

node_modules/@grpc/grpc-js/build/src/server-call.d.ts:77:5 - error TS2416: Property 'end' in type 'ServerDuplexStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Duplex'.
  Type '(metadata?: any) => void' is not assignable to type '{ (cb?: () => void): this; (chunk: any, cb?: () => void): this; (chunk: any, encoding?: BufferEncoding, cb?: () => void): this; }'.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

77     end(metadata?: any): void;
       ~~~


Found 3 errors.
```
- `yarn add @types/node@17.0.5 && tsc -b`
- successfully built again.

### Environment
 - OS name, version and architecture: M1 Pro, Macbook Pro
 - Node version: 16.13.0
 - Node installation method: brew
 - Package name and version: `"@grpc/grpc-js": "^1.4.5"`

### tsconfig.json
`{
  "compilerOptions": {
    "lib": [
      "ES2021"
    ],
    "module": "commonjs",
    "moduleResolution": "node",
    "target": "ES2021",
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "outDir": "dist"
  },
  "include": [
    "src/**/*"
  ],
  "exclude": [
    "node_modules",
    "**/*.spec.ts"
  ]
}`

It seems that the version of dependency package `@types/node` should be limited, rather than current ">=12.12.47".

**Comment by beerent** [comment]
thank you

**Comment by johnnyoshika** [comment]
I'm getting the same error but my `@types/node` is version 14

```
"@types/node": "^14.14.22"
```

`@grpc/grpc-js@1.4.5` is a dependency of `@google-cloud/compute@3.0.0` in my project:

![image](https://user-images.githubusercontent.com/504505/147867858-74b652c9-3a8d-4ce4-a464-2847aef192b8.png)

**Comment by wiegell** [comment]
@murgatroid99 sorry to ping you, but this issue has traction - seems to be affecting a lot of user

**Comment by anteAdamovic** [comment]
Same issue for me, buy my @types/node is 14.17.0 so it doesn't feel like it's an issue with node types themselves ... Error I'm getting is:

```
remote:        node_modules/@grpc/grpc-js/build/src/object-stream.d.ts(11,18): error TS2430: Interface 'IntermediateObjectWritable<T>' incorrectly extends interface 'Writable'.
remote:          The types returned by 'end(...)' are incompatible between these types.
remote:            Type 'void' is not assignable to type 'this'.
remote:              'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.
remote:        node_modules/@grpc/grpc-js/build/src/server-call.d.ts(64,5): error TS2416: Property 'end' in type 'ServerWritableStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Writable'.
remote:          Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding: BufferEncoding, cb?: (() => void) | undefined): this; }'.
remote:            Type 'void' is not assignable to type 'this'.
remote:              'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.
remote:        node_modules/@grpc/grpc-js/build/src/server-call.d.ts(77,5): error TS2416: Property 'end' in type 'ServerDuplexStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Duplex'.
remote:          Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding?: BufferEncoding | undefined, cb?: (() => void) | undefined): this; }'.
remote:            Type 'void' is not assignable to type 'this'.
remote:              'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.
```

Still working on a fix ... unable to deploy production app.

**Comment by Net301** [comment]
We are experiencing the same error. we are unable to deploy to production, is there any estimation of when the patch will be published ?

**Comment by darioxtx** [comment]
It seems like `@types/node` definition was changed. Using `"@types/node": "14.18.3"` build works fine.

**Comment by russelldias98** [comment]
> It seems like `@types/node` definition was changed. Using `"@types/node": "14.18.3"` build works fine.

This worked for me! Thanks @darioxtx

**Comment by murgatroid99** [comment]
I will fix this when I am back to work on Tuesday. Until then, it looks like there is an effective workaround.

> It seems that the version of dependency package `@types/node` should be limited, rather than current ">=12.12.47".

Unfortunately, because of how TypeScript does module resolution, that causes other more persistent problems, detailed in #1679.

**Comment by johnnyoshika** [comment]
Interesting. It does seem to be a problem with `@types/node@14.18.4`. I just created a standalone test project to troubleshoot this. By default, `@types/node@14.18.4` was used. This was my default output for `npm ls @types/node`:

```
├─┬ @google-cloud/compute@3.0.0
│ └─┬ google-gax@2.28.1
│   ├─┬ @grpc/grpc-js@1.4.5
│   │ └── @types/node@14.18.4 deduped
│   └─┬ protobufjs@6.11.2
│     └── @types/node@14.18.4 deduped
├─┬ @types/express@4.17.13
│ ├─┬ @types/body-parser@1.19.2
│ │ ├─┬ @types/connect@3.4.35
│ │ │ └── @types/node@14.18.4 deduped
│ │ └── @types/node@14.18.4 deduped
│ ├─┬ @types/express-serve-static-core@4.17.27
│ │ └── @types/node@14.18.4 deduped
│ └─┬ @types/serve-static@1.13.10
│   └── @types/node@14.18.4 deduped
├── @types/node@14.18.4
└─┬ firebase-admin@9.12.0
  └── @types/node@14.18.4 deduped
```

But when I tried to compile with `tsc`, I got the error described in this issue:

```
> tsc

  The types returned by 'end(...)' are incompatible between these types.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

11 export interface IntermediateObjectWritable<T> extends Writable {
                    ~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@grpc/grpc-js/build/src/server-call.d.ts:64:5 - error TS2416: Property 'end' in type 'ServerWritableStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Writable'.
  Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding: BufferEncoding, cb?: (() => void) | undefined): this; }'.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

64     end(metadata?: any): void;
       ~~~

  Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding?: BufferEncoding | undefined, cb?: (() => void) | undefined): this; }'.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

77     end(metadata?: any): void;
       ~~~


Found 3 errors.
```

But now when I force `@types/node@14.18.3` as @darioxtx suggested, everything works. Here's the output of `npm ls @types/node` after the fix:

```
├─┬ @google-cloud/compute@3.0.0
│ └─┬ google-gax@2.28.1
│   ├─┬ @grpc/grpc-js@1.4.5
│   │ └── @types/node@14.18.3 deduped
│   └─┬ protobufjs@6.11.2
│     └── @types/node@14.18.3 deduped
├─┬ @types/express@4.17.13
│ ├─┬ @types/body-parser@1.19.2
│ │ ├─┬ @types/connect@3.4.35
│ │ │ └── @types/node@14.18.3 deduped
│ │ └── @types/node@14.18.3 deduped
│ ├─┬ @types/express-serve-static-core@4.17.27
│ │ └── @types/node@14.18.3 deduped
│ └─┬ @types/serve-static@1.13.10
│   └── @types/node@14.18.3 deduped
├── @types/node@14.18.3
└─┬ firebase-admin@9.12.0
  └── @types/node@14.18.3 deduped
```

**Comment by dancamdev** [comment]
Ok, looks like I'm not alone on this issue.

@johnnyoshika how do you force the `@types/node@14.18.3` version?

it doesn't seem to be working right for me.

**Comment by johnnyoshika** [comment]
@dancamdev Just change `@types/node` version in `package.json` to this:

```
"@types/node": "14.18.3"
```

Then:

```
npm install
```

**Comment by dancamdev** [comment]
@johnnyoshika thanks! the one in the grpc module? because changing it in my project one doesn't do the trick apparently.

**Comment by johnnyoshika** [comment]
@dancamdev No, you should only be changing `package.json` in your project. What is the output of `npm ls @types/node`?

**Comment by dancamdev** [comment]
@johnnyoshika Fixed that, thank you!

**Comment by dancamdev** [comment]
I'm not sure it's related but once I changed the `@types/node` version I'm getting this error in the logs and aren't able to deploy the functions anyway.

`Could not find image for function projects/path/to/function`

Here's my `package.json` file:

```
{
  "name": "functions",
  "scripts": {
    "lint": "eslint --ext .js,.ts .",
    "build": "tsc",
    "serve": "npm run build && firebase emulators:start --only functions",
    "shell": "npm run build && firebase functions:shell",
    "start": "npm run shell",
    "deploy": "firebase deploy --only functions",
    "logs": "firebase functions:log"
  },
  "engines": {
    "node": "16"
  },
  "main": "lib/index.js",
  "dependencies": {
    "firebase-admin": "^9.8.0",
    "firebase-functions": "^3.14.1"
  },
  "devDependencies": {
    "@typescript-eslint/eslint-plugin": "^3.9.1",
    "@typescript-eslint/parser": "^3.8.0",
    "eslint": "^7.6.0",
    "eslint-config-google": "^0.14.0",
    "eslint-plugin-import": "^2.22.0",
    "firebase-functions-test": "^0.2.0",
    "typescript": "^3.8.0",
    "@types/node": "14.18.3"
  },
  "private": true
}
```

**Comment by oceanroleplay** [comment]
if it helps, enable ``skipLibCheck`` in your `tsconfig.json` to build project until this issue resolved.

**Comment by natarajinnavoto** [comment]
I am getting the below error 

node_modules/@grpc/grpc-js/build/src/object-stream.d.ts:11:18 - error TS2430: Interface 'IntermediateObjectWritable<T>' incorrectly extends interface 'Writable'.
  The types returned by 'end(...)' are incompatible between these types.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

11 export interface IntermediateObjectWritable<T> extends Writable {
                    ~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@grpc/grpc-js/build/src/server-call.d.ts:64:5 - error TS2416: Property 'end' in type 'ServerWritableStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Writable'.
  Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding: BufferEncoding, cb?: (() => void) | undefined): this; }'.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

64     end(metadata?: any): void;
       ~~~

node_modules/@grpc/grpc-js/build/src/server-call.d.ts:77:5 - error TS2416: Property 'end' in type 'ServerDuplexStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Duplex'.
  Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding?: "ascii" | "utf8" | "utf-8" | "utf16le" | "ucs2" | "ucs-2" | "base64" | "base64url" | "latin1" | "binary" | "hex" | undefined, cb?: (() => void) | undefined): this; }'.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

77     end(metadata?: any): void;
       ~~~
Found 3 errors.
Error: functions predeploy error: Command terminated with non-zero exit code2


npm ls @types/node

functions@ ....

+-- firebase-admin@10.0.1
| +-- @google-cloud/firestore@4.15.1
| | +-- google-gax@2.28.1
| | | `-- @grpc/grpc-js@1.4.5
| | |   `-- @types/node@17.0.6 deduped
| | `-- protobufjs@6.11.2
| |   `-- @types/node@17.0.6 deduped
| `-- @types/node@17.0.6
`-- firebase-functions@3.16.0
  `-- @types/express@4.17.3
    +-- @types/body-parser@1.19.2
    | +-- @types/connect@3.4.35
    | | `-- @types/node@17.0.6 deduped
    | `-- @types/node@17.0.6 deduped
    +-- @types/express-serve-static-core@4.17.27
    | `-- @types/node@17.0.6 deduped
    `-- @types/serve-static@1.13.10
      `-- @types/node@17.0.6 deduped

**Comment by julian-hecker** [comment]
This is also affecting me in from firebase admin. I found that using @types/node@14 works as a decent workaround.

**Comment by johnnyoshika** [comment]
@julian-hecker I use `firebase-admin` as well. So far I haven't had any issues. What breaks for you?

**Comment by paxforce** [comment]
Hi @johnnyoshika - I changed the @types/node to 14.18.3 as you suggested, but I'm still getting the error every time I attemtp to `npm run build`
```
node_modules/@grpc/grpc-js/build/src/object-stream.d.ts:11:18 - error TS2430: Interface 'IntermediateObjectWritable<T>' incorrectly extends interface 'Writable'.
  The types returned by 'end(...)' are incompatible between these types.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

11 export interface IntermediateObjectWritable<T> extends Writable {
                    ~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@grpc/grpc-js/build/src/server-call.d.ts:64:5 - error TS2416: Property 'end' in type 'ServerWritableStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Writable'.
  Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding: BufferEncoding, cb?: (() => void) | undefined): this; }'.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

64     end(metadata?: any): void;
       ~~~

node_modules/@grpc/grpc-js/build/src/server-call.d.ts:77:5 - error TS2416: Property 'end' in type 'ServerDuplexStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Duplex'.
  Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding?: "ascii" | "utf8" | "utf-8" | "utf16le" | "ucs2" | "ucs-2" | "base64" | "base64url" | "latin1" | "binary" | "hex" | undefined, cb?: (() => void) | undefined): this; }'.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

77     end(metadata?: any): void;
       ~~~

src/index.ts:1:1 - error TS6133: 'functions' is declared but its value is never read.

1 import * as functions from "firebase-functions";
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 4 errors.

npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! functions@ build: `tsc`
npm ERR! Exit status 2
```

**Comment by wiegell** [comment]
> Hi @johnnyoshika - I changed the @types/node to 14.18.3 as you suggested, but I'm still getting the error every time I attemtp to `npm run build` (after that I did the `yarn install`, of course)
> 
> ```
> node_modules/@grpc/grpc-js/build/src/object-stream.d.ts:11:18 - error TS2430: Interface 'IntermediateObjectWritable<T>' incorrectly extends interface 'Writable'.
>   The types returned by 'end(...)' are incompatible between these types.
>     Type 'void' is not assignable to type 'this'.
>       'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.
> 
> 11 export interface IntermediateObjectWritable<T> extends Writable {
>                     ~~~~~~~~~~~~~~~~~~~~~~~~~~
> 
> node_modules/@grpc/grpc-js/build/src/server-call.d.ts:64:5 - error TS2416: Property 'end' in type 'ServerWritableStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Writable'.
>   Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding: BufferEncoding, cb?: (() => void) | undefined): this; }'.
>     Type 'void' is not assignable to type 'this'.
>       'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.
> 
> 64     end(metadata?: any): void;
>        ~~~
> 
> node_modules/@grpc/grpc-js/build/src/server-call.d.ts:77:5 - error TS2416: Property 'end' in type 'ServerDuplexStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Duplex'.
>   Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding?: "ascii" | "utf8" | "utf-8" | "utf16le" | "ucs2" | "ucs-2" | "base64" | "base64url" | "latin1" | "binary" | "hex" | undefined, cb?: (() => void) | undefined): this; }'.
>     Type 'void' is not assignable to type 'this'.
>       'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.
> 
> 77     end(metadata?: any): void;
>        ~~~
> 
> src/index.ts:1:1 - error TS6133: 'functions' is declared but its value is never read.
> 
> 1 import * as functions from "firebase-functions";
>   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 
> 
> Found 4 errors.
> 
> npm ERR! code ELIFECYCLE
> npm ERR! errno 2
> npm ERR! functions@ build: `tsc`
> npm ERR! Exit status 2
> ```

Are you using both yarn and npm? When using yarn 1, I still had the non-working 17.0.6 types in the grpc module even after changing to 14x in the main package. Probably bc. of defective dedupe in yarn1. Try and see yourself with ‘yarn list —pattern @types/node‘.
I solved it by finally upgrading to yarn3.

**Comment by grascioni** [comment]
Just my two cents, not sure if this is the right solution though.

I had another project that kept compiling without issues, and I noticed its package-lock file had the following version specified in it:

>     "@types/node": {
>          "version": "16.11.11",
>          "resolved": "https://registry.npmjs.org/@types/node/-/node-16.11.11.tgz",
>          "integrity": "sha512->    KB0sixD67CeecHC33MYn+eYARkqTheIRNuu97y2XMjR7Wu3XibO1vaY6VBV6O/a89SPI81cEUIYT87UqUWlZNw=="
>     },

So I just went ahead and removed the package-lock.json line from the gitignore file in my broken project and overwrote the @types/node version with the above one.

**Comment by johnnyoshika** [comment]
@paxforce I don't use `yarn` so I'm not sure why you're encountering this problem. What is the output of `npm ls @types/node` (or the equivalent yarn command)?

**Comment by anilnabin25** [comment]
> It seems like `@types/node` definition was changed. Using `"@types/node": "14.18.3"` build works fine.

thankx this solved my problem but used `"@types/node": "17.0.5"`another version

**Comment by paxforce** [comment]
@johnnyoshika the output of ls @types/node is the following
```
+-- firebase-admin@9.12.0
| +-- @google-cloud/firestore@4.15.1
| | +-- google-gax@2.28.1
| | | `-- @grpc/grpc-js@1.4.5
| | |   `-- @types/node@17.0.7  deduped
| | `-- protobufjs@6.11.2
| |   `-- @types/node@17.0.7  deduped
| `-- @types/node@17.0.7
`-- firebase-functions@3.16.0
  `-- @types/express@4.17.3
    +-- @types/body-parser@1.19.2
    | +-- @types/connect@3.4.35
    | | `-- @types/node@17.0.7  deduped
    | `-- @types/node@17.0.7  deduped
    +-- @types/express-serve-static-core@4.17.27
    | `-- @types/node@17.0.7  deduped
    `-- @types/serve-static@1.13.10
      `-- @types/node@17.0.7  deduped
```

Perhaps it's also worth mentioning that it's a project inside an Nx mono-repo.
I added firebase-related stuff to the project (to the project's package.json)
"firebase": "^9.6.1",
"firebase-admin": "^10.0.1",
"@angular/fire": "^7.2.0",
- I initialized firebase in my project `firebase init` - this among other things, created a `functions` folder with it's own package.json file.
- now inside the `functions` folder I tried to install dependencies (yarn install), however this didn't work because it expected node 14 and I have node 16 installed globally. So I had to add the ignore engines flag (yarn install --ignore-engines) and that one worked - it created the functions/node_modules folder with deps.
- however, when I'm trying to build the functions project, the `yarn run build` command doesn't work even with the --ignore-engines flag.
```
error functions@: The engine "node" is incompatible with this module. Expected version "14". Got "16.13.1"
error Commands cannot run with an incompatible environment
```

**Comment by johnnyoshika** [comment]
@paxforce I can't speak to the node incompatibility issue, but I also run my project inside of a monorepo (using npm workspaces). It looks like you're using `@types/node@17.0.7`, which based on @anilnabin25's comment should work. I'm stumped why you're still encountering the `@types/node` problem.

**Comment by paxforce** [comment]
@johnnyoshika would you be so kind and look at my repo? I think if you looked at the code you could tell what's wrong with it. I've been stuck with this issue for 3 days now :( ...

**Comment by murgatroid99** [comment]
Thank you for your patience with this error. I have published `@grpc/grpc-js` version 1.4.6 with a fix for it. Please update your dependencies to use the new version.
